[ty] Emit diagnostic for `NamedTuple` and `TypedDict` decorated with dataclass (#22672)
astral-sh/ruff: crates/ty_python_semantic/resources/mdtest/typed_dict.md

        pass
```

## `TypedDict` with `@dataclass` decorator

Applying `@dataclass` to a `TypedDict` class is invalid and will cause an `AttributeError` at

> runtime when instantiating the class:

Reviewer: I'm not sure it'll lead to an `AttributeError`, but it can definitely lead to other kinds of errors. Whether it does depends on exactly how you try to instantiate it:

```pycon
>>> from typing import TypedDict
>>> from dataclasses import dataclass
>>> @dataclass
... class Foo(TypedDict):
...     x: str
...     
>>> Foo(x='x')
{'x': 'x'}
>>> Foo('x')
Traceback (most recent call last):
  File "<python-input-6>", line 1, in <module>
    Foo('x')
    ~~~^^^^^
ValueError: dictionary update sequence element #0 has length 1; 2 is required
```

I think the better rationale when it comes to `TypedDict`s is possibly just that it's conceptually incoherent to apply `@dataclass` to a `TypedDict` class -- `TypedDict`s are abstract structural types that have no effect on the runtime class ("instantiating" a `TypedDict` always gives you a `dict` at runtime), whereas `@dataclass` is a tool for easily customising the creation of new nominal types/classes at runtime

        pass
```

## `TypedDict` with `@dataclass` decorator

Applying `@dataclass` to a `TypedDict` class is conceptually incoherent: `TypedDict` defines
abstract structural types where "instantiating" always gives you a plain `dict` at runtime, whereas

> `@dataclass` is a tool for customising the creation of new nominal types. An exception will be

Reviewer: ```suggestion
`@dataclass` is a tool for customising the creation of new nominal types. An exception may be
```

raised when instantiating the class at runtime:

```py
from dataclasses import dataclass
from typing import TypedDict

@dataclass
# error: [invalid-dataclass] "`TypedDict` class `Foo` cannot be decorated with `@dataclass`"
class Foo(TypedDict):
    x: int
    y: str
```

> The same error occurs with `dataclasses.dataclass` used as a function call:

Reviewer: ```suggestion
The same error occurs with `dataclasses.dataclass` used with parentheses:
```
